### ISSUE TITLE
Questions about BlenderBot optimization

### ISSUE THREAD

**Issue by g-karthik** [facebookresearch/ParlAI#3488]
The Blender paper says: "Both our 2.7B and 9.4B parameter models were trained with batches of approximately 500k label BPE tokens per batch. The 2.7B parameter model trained for approximately 200k SGD updates with a maximum learning rate of 2e-4, a linear warmup of 3125 steps, and an invsqrt LR scheduler (Vaswani et al., 2017); the model had not converged when we stopped. The 9.4B parameter model was trained with a maximum learning rate of 1.15e-4 and 2400 warmup steps for a total of 200k SGD updates, and did not appear to be overfitting."

What was the actual batch size per GPU, what was the maximum sequence length ("500k label BPE tokens per batch" is uninformative)? How many GPUs were actually used in total?

What was the number of steps for gradient accumulation? By 200k SGD updates, do you mean the `optimizer.step()` was called 200k times? Or do you mean 200k iterations happened but the actual number of "steps"/param updates was 200k / N, where N is the number of gradient accumulation steps?

**Comment by g-karthik** [comment]
Also, in the start of the same paragraph:

"We perform pre-training using the Fairseq (Ott et al., 2019) toolkit. Our 2.7B and 9.4B parameter models were both trained using the Adam optimizer (Kingma and Ba, 2014). In order to fit the larger models onto nodes, we utilize Megatron-LM style model parallelism (Shoeybi et al., 2019), in which the Feed Forward network (FFN) and Multihead Attention layers of the Transformer are “vertically” sliced"

I'm quite confused - first it says Adam, then it says 200k SGD updates. What was the optimizer used?

**Comment by stephenroller** [comment]
- Maximum sequence length was 128 context, 128 label. Average much lower though, we packed batches, so the batch size was dynamic throughout training. Something like 128 GPUs was used for the 2.7B if I recall correctly, but we might've upped that to 256 after some point to have a lower gradient accumulation.
- 200k sgd steps means _after accumulation_. Meaning 200k calls to optimizer.step()
- Adam is also a stochastic gradient descent optimizer.

**Comment by g-karthik** [comment]
@stephenroller Thanks for the quick reply!

When you say "we packed batches, so the batch size was dynamic throughout training", are you alluding to dynamic padding via `collate_fn` causing the actual number of tokens in a batch being dynamic throughout training? I tend to think of batch size as the number of examples, not the number of tokens in the batch. So it'd also help to know the number of examples per GPU (unless that was the one that you alluded to as dynamic, in which case I am happy to learn something new about PyTorch's capabilities from you today).

Also, can you share the number of gradient accumulation steps performed before each Adam update?

It would be easier if you could just paste the fairseq CLI arguments that were used.

**Comment by stephenroller** [comment]
The number of gradient accumulation steps was probably 2 or 4 depending on how many gpus we were using. These things are interchangeable, as they both just increase effective batchsize.

Yes, dynamic padding is worth something like a 2-3x speedup, and is supported both in fairest and parlai. The idea is that I can have a batch with a handful of very long utterances, or I can have a batch with many, many short utterances. This helps maximize your GPU memory and utilization.

Back of the envelope I guess we probably got about 32 examples per gpu per optimizer step on average, and that was likely accomplished via 128 gpus and update freq 4, or 256 gpus and update freq 2.

Usually when you use this many gpus, gradient accumulation speeds things up by reducing communication, which tends to be very expensive. These days you could do better using fully sharded data parallel, probably getting by with only 32 gpus or so.

**Comment by kazeto111** [comment]
@stephenroller 
I couldn't believe 256GPUs!
How long did it take to train 2.7B with 256GPUs?

Sorry for one more question.
When I used trained 2.7B model with blended_skill_talk task, it spends 10 second per 1 response with 4 Tesla T4 GPUs despite of 3 second with 1 GPU. So I thought 256GPUs takes more time than fewer GPUs( I thought distributed processing takes much time). How did you use GPUs?
